### PR TITLE
[M1573] Support null Quadrat benthic percent values

### PR DESCRIPTION
**Pull request by gridcell**

### CODE REVIEW

**Comment by gridcell** [comment]
[allow nulls for bleaching QC percent cover observations](https://trello.com/c/QlBm2v0h/1573-allow-nulls-for-bleaching-qc-percent-cover-observations)

File: src/api/submission/validations.py [hunk 1]
```diff
@@ -1075,56 +1075,58 @@ def get_observations(cls, data, key=None):
...
+                return self.error(self.identifier, self.LESS_QUADRAT_NUMBER)
 
-            if (
-                self._validate_percent_value_total(values, "validate_percent_values")
-                == ERROR
-            ):
-                return ERROR
+            quadrat_nums.append(quadrat_num)
+
+        if len(quadrat_nums) != len(set(quadrat_nums)):
```
Comment (saanobhaai) [inline src/api/submission/validations.py:1128]: This tests that the number of quadrat numbers is the same -- but couldn't they be the same and still have a duplicate? I'm probably not seeing the whole logic.

File: src/api/resources/obs_quadrat_benthic_percent.py [hunk 2]
```diff
@@ -1,13 +1,28 @@
...
-        extra_kwargs = {}
+        extra_kwargs = {
+            "quadrat_number": {
+                "error_messages": {
+                    "null": "Quadrat number is required"
+                }
+            }
+        }
+        validators = [
+            UniqueTogetherValidator(
```
Comment (saanobhaai) [inline src/api/resources/obs_quadrat_benthic_percent.py:20]: Might be good to talk this through... not quite getting why we need a special serializer validator in addition to the submission validations (and/or model `unique_together`), or how they interact...

File: src/api/submission/validations.py [hunk 3]
```diff
@@ -1075,56 +1075,58 @@ def get_observations(cls, data, key=None):
...
     VALID_NUMBER_MSG = "Percent cover value not between 0 and 100"
+    LESS_QUADRAT_NUMBER = "Quadrat number less than 1"
+    DUPLICATE_QUADRAT_NUMBERS = "Duplicate quadrat numbers"
 
-    def _check_percent_value(self, value, msg_param):
-        try:
-            _ = int(value)
-        except (TypeError, ValueError):
-            return ERROR, self.VALID_NUMBER_MSG
+    def validate_percent_values(self):
```
Comment (saanobhaai) [inline src/api/submission/validations.py:1085]: OK I think I get this, and it's much better...

File: src/api/submission/validations.py (hunk 1 above)
Comment (gridcell) [reply src/api/submission/validations.py:1128]: it's checking the length of the list versus the length of the set of quadrat_nums ?

File: src/api/resources/obs_quadrat_benthic_percent.py (hunk 2 above)
Comment (gridcell) [reply src/api/resources/obs_quadrat_benthic_percent.py:20]: Custom error message wasn't working.  Apparently there was a bug related to this in an earlier version but the version we are currently running should have this fix in it.

**Pull request merged**